File: onion-architecture/java/src/main/java/com/example/application/EnrolmentApplicationService.java
package com.example.application;

import com.example.application.ports.CourseRepository;
import com.example.application.ports.StudentRepository;
import com.example.domain.Course;
import com.example.domain.Student;
import com.example.domain.services.EnrolmentService;

public class EnrolmentApplicationService {

    private final StudentRepository studentRepo;
    private final CourseRepository courseRepo;
    private final EnrolmentService enrolmentService;

    public EnrolmentApplicationService(
            StudentRepository studentRepo,
            CourseRepository courseRepo,
            EnrolmentService enrolmentService
    ) {
        this.studentRepo = studentRepo;
        this.courseRepo = courseRepo;
        this.enrolmentService = enrolmentService;
    }

    public void enrol(Long studentId, Long courseId) {
        Student student = studentRepo.findById(studentId)
                .orElseThrow(() -> new RuntimeException("Student not found"));
        Course course = courseRepo.findById(courseId)
                .orElseThrow(() -> new RuntimeException("Course not found"));

        enrolmentService.enrolStudent(student, course);
        studentRepo.save(student); // save updated enrolment
    }
}

File: onion-architecture/kotlin/src/main/kotlin/com/example/application/EnrolmentApplicationService.kt
package com.example.application

import com.example.application.ports.CourseRepository
import com.example.application.ports.StudentRepository
import com.example.domain.services.EnrolmentService

class EnrolmentApplicationService(
    private val studentRepo: StudentRepository,
    private val courseRepo: CourseRepository,
    private val enrolmentService: EnrolmentService
) {
    fun enrol(studentId: Long, courseId: Long) {
        val student = studentRepo.findById(studentId)
            ?: throw RuntimeException("Student not found")
            
        val course = courseRepo.findById(courseId)
            ?: throw RuntimeException("Course not found")

        enrolmentService.enrolStudent(student, course)
        studentRepo.save(student) // save updated enrolment
    }
}